libopae/api-shell.c

```diff
@@ -218,19 +222,34 @@ fpga_result fpgaGetPropertiesFromHandle(fpga_handle handle,
 
 	// If the output properties has a parent token set,
 	// then it will be a raw token. We need to wrap it.
-	if (fpgaPropertiesGetParent(*prop, &parent) == FPGA_OK) {
-		opae_wrapped_token *wrapped_parent =
-			opae_allocate_wrapped_token(
-				parent, wrapped_handle->adapter_table);
 
-		if (!wrapped_parent) {
-			OPAE_ERR("malloc failed");
-			return FPGA_NO_MEMORY;
-		}
+	p = opae_validate_and_lock_properties(*prop);
+
+	ASSERT_NOT_NULL(p);
+
+	if (FIELD_VALID(p, FPGA_PROPERTY_PARENT)) {
+		fpga_token parent = NULL;
+
+		res = wrapped_handle->adapter_table->fpgaCloneToken(
+			p->parent, &parent);
 
-		return fpgaPropertiesSetParent(*prop, wrapped_parent);
+		if (res == FPGA_OK) {
+			opae_wrapped_token *wrapped_parent =
+				opae_allocate_wrapped_token(
+					parent, wrapped_handle->adapter_table);
+
+			if (wrapped_parent) {
+				p->parent = wrapped_parent;
```

Comment:
If we clone p->parent, and then assign p->parent to wrapped_parent, what happens to the original p->parent? Isn't that leaked?

Reply:
I wanted to talk to you about this.  I'm following the precedent that was established in the properties API.  Look at fpgaPropertiesGetParent().  You'll notice that it is cloning the parent token and handing back the clone.  I question whether we need to do that in either place actually.  Why is fpgaPropertiesGetParent() cloning the token?